README.md
# jac : just another cache

jac is a persistent in-memory key:value/store cache inspired by go-cache that is basically am augmented thread-safe collections of `map[string]string` with expiration time called buckets.  
jac is suitable for application running on a single machine that need persistence and resistance towards system or application crashes. It achieved this by duplicating the data itself on the disk in a readable manner (optional) during execution and as a standard GOB once the application has closed. In this way it is possible to load a the latest cache state after an application restart or a system crash.  
jac generates two files either during operation (.data) or for shutdown back-up (.rec). The former is a JSON text file that is sued to reload the state in case of sudden system or application death. It can also be used to preload values into the cache.  The latter are GOB save files generated when the cache is properly closed and are the rpeferred mode to load the cache at application (re)start.  

jac has been compiled and tested for the following pairs:  

 - linux/arm  
 - linux/arm64  
 - linux/amd64  
 - openWRT/mipsle 
 - windows/amd64  

jac is built to accept data of any type, however the data will be always stored as a string and, to be type agnostic, will be also returned as a string.  

### Installation

`go get github.com/fpessolano/jac`

### Usage

```go
package main

import (
	"fmt"
	"github.com/fpessolano/jac"
	"os"
)

func main() {
    // Initialise the cache. The options below are the standard values
    if err := jac.Initialise(true, nil); err != jac.IllegalParameter && err != nil {
        fmt.Println(err)
        os.Exit(0)
    }
	
    // Create a cache bucket called test with a default expiration time (see types.go)
    test, err := jac.NewBucket("definitions", jac.NoExpiration)
    if err != nil {
        fmt.Println(err)
        os.Exit(0)
    }

    // Set a persistent value of the key "one" to "1", with the default expiration time
    test.Set("one", 1, jac.DefaultExpiration, true)
    
    // Set a non-persistent value of the key "b" to "2", with no expiration time
    test.Set("two", "two", jac.NoExpiration, false)
    
    // Get the value associated with the key "one" from the cache
    oneRead, found := test.Get("two")
    if found {
        fmt.Println(oneRead)
    }
	
    // Get all items in bucket test
    allItems := test.Items()
    for _, el := range(allItems) {
        fmt.Println(el)	
    }
    
    // Close the bucket
    test.Close(false)
    
    // Close the cache
    jac.Terminate()
}
```

### Available methods

```go
    // Initialise prepare the cache for use. It accept two parameters:
    //  v : set to true for verbose (useful for development purposes)
    //  o : are the database options (see inm types.go for further details)
    func Initialise(v bool, o *Options) error
    
    // Terminate closes the cache and relative processes
    func Terminate()
    
    // NewBucket create a new bucket in the cache.
    //  If a rec file or a data file are present and are not older than time.Now() - maxage,
    //  they will be loaded in the cache
    func NewBucket(name string, exp time.Duration) (c Bucket, e error) 
    
    // Close closes a bucket storing values in the recovery data
    //  if keep is false the working data file will be deleted
    func (c *Bucket) Close(keep bool) 
    
    // Get read the value associated to the key k/
    //  It also returns false if the key has not value associated to it
    func (c *Bucket) Get(k string) (string, bool) 
    
    // GetWithExpiration performs the same operation as Get but it also returns the
    //  value expiration time
    func (c *Bucket) GetWithExpiration(k string) (string, time.Time, bool) 
    
    // Set writes a new key/value pair with a given expiration time t and
    //  It marks the key/value pair persistent if pers is true.
    func (c *Bucket) Set(k string, vn interface{}, t time.Duration, pers bool) 
    
    // Update updates the key/value pair with a given expiration time t and
    //  It marks the key/value pair persistent if pers is true. The working file is not changed
    //  even if pers is true
    func (c *Bucket) Update(k string, vn interface{}, t time.Duration, pers bool) 
    
    // Replace replaces an existing key/value pair only it already existing.
    //  It marks the key/value pair persistent if pers is true.
    func (c *Bucket) Replace(k string, vn interface{}, t time.Duration, pers bool) 
    
    // Add add a new key/value pair only if it does not already
    //  It marks the key/value pair persistent if pers is true.
    func (c *Bucket) Add(k string, vn interface{}, t time.Duration, pers bool) (string, bool) 
    
    // FunctionUpdate updates the new key/value with a custom function of type
    //  func(k, v string) (string, string). The function is given the actual key/value pair is present
    //  and key/"" otherwise and it expects a new key/value pair.
    //  if the function modofies the key, the old key/pair is deleted.
    //  It marks the key/value pair persistent if pers is true.
    func (c *Bucket) FunctionUpdate(k string, f updateFunc, t time.Duration, pers bool) (string, string, bool) 
    
    // Items returns all elements in the bucket as a map[string]string
    func (c *Bucket) Items() (rt map[string]string) 
    
    // Delete permanently removes an item from the bucket
    func (c *Bucket) Delete(k string) 
    
    // Compact initiate a compation request of the working files.
    //  Please note that all values will be used and any previously indicated persistence flag will be ignored.
    func (c *Bucket) Compact() 
    
    // Flush deletes all items from the bucket.
    func (c *Bucket) Flush() 
    
    // ItemCount returns the number of items in the bucket. This may include items that have
    //  expired, but have not yet been cleaned up.
    func (c *Bucket) ItemCount() int 
    
    // OnEvicted sets an (optional) function that is called with the key and value when an
    //  item is evicted from the bucketInternal. (Including when it is deleted manually, but
    //  not when it is overwritten.) Set to nil to disable.
    func (c *Bucket) OnEvicted(f func(string, interface{})) 
    
    // DeleteExpired deletes all expired items from the bucketInternal.
    func (c *Bucket) DeleteExpired() 
```



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.1]
### Fixed  
 - README example has been fixed
 - go.mod added  

## [0.2.0]
### Changed  
 - Set accept any value, however get always return a string.   


## [0.1.0]
### This is the first released version with all initial functionality




jac_test.go
package jac

import (
	"encoding/json"
	"fmt"
	"strconv"
	"testing"
	"time"
)

type testData struct {
	Id    string `json:"id"`
	Flag1 bool   `json:"flag1"`
	Value int    `json:"value"`
	Flag2 bool   `json:"flag2"`
}

//goland:noinspection GoNilness
func Test_jac(t *testing.T) {
	const (
		Tries = 100
		Keep  = true
	)

	if err := Initialise(true, nil); err != IllegalParameter && err != nil {
		t.Errorf(err.Error())
	}
	println("create")
	bucket, e := NewBucket("test", 180*time.Second)
	if e != nil {
		t.Errorf(e.Error())
	}
	for i := 0; i < Tries; i++ {
		tmp := testData{
			Id:    strconv.Itoa(i),
			Flag1: false,
			Value: i,
			Flag2: false,
		}
		if msg, er := json.Marshal(tmp); er == nil {
			bucket.Set(strconv.Itoa(i), string(msg), 180*time.Second, true)
		}
		time.Sleep(50 * time.Millisecond)
	}
	time.Sleep(2 * time.Second)
	bucket.Close(true)

	time.Sleep(2 * time.Second)
	println("verify")
	bucket, e = NewBucket("test", 180*time.Second)
	if e != nil {
		t.Errorf(e.Error())
	}
	for i := 0; i < Tries; i++ {
		msg, found := bucket.Get(strconv.Itoa(i))
		if !found {
			t.Errorf("missing data")
		} else {
			var entry testData
			if err := json.Unmarshal([]byte(msg), &entry); err == nil {
				fmt.Println(entry)
				if entry.Value != i {
					t.Errorf("wrong data")
				}
			}
		}
	}
	time.Sleep(90 * time.Second)
	bucket.Close(Keep)

	Terminate()
	time.Sleep(2 * time.Second)

}


types.go
package jac

import (
	"os"
	"sync"
	"time"
)

// external data types

// Default options are
//	ExpirationTime:     0,
//	IntervalCompacting: 1440 * 60,
//	InternalBuffering:  10,
//	LoadDelayMs:        10,
//	MaximumAge:         5 * 60,
type Options struct {
	ExpirationTime     int    // Expiration time is seconds
	IntervalCompacting int    // Cache working files compacting interval in seconds (values smaller than 60s will be defaulted to 60s)
	InternalBuffering  int    // Buffering length to decouple the in-memory cache from the disk processes. Bigger numbers improve cache speed at expenses of system crash resistance
	LoadDelayMs        int    // Regulates the start-up delay. Smaller numbers improves start-up time at costs of possible loss of persistence
	MaximumAge         int64  // Maximum age (in s) of a back-up file (.rec) or working file (.data) for it to be used to initialise the cache
	WorkingFolder      string // folder for working files (.data). File contain the entire cache in a readable. Altering the files only affects the initial cache load not its operation
	RecoveryFolder     string // folder for back-up files (.rec)
}

type Item struct {
	Object     interface{}
	Expiration int64
}

type FileData struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

type Bucket struct {
	name   string
	bucket *bucketInternalPtr
	file   *os.File
	writer chan backupData
	cr     chan interface{}
}

// internal data types
type bucketInternalPtr struct {
	*bucketInternal
}

type bucketInternal struct {
	defaultExpiration time.Duration
	items             map[string]Item
	mu                sync.RWMutex
	onEvicted         func(string, interface{})
	janitor           *janitor
}

type keyAndValue struct {
	key   string
	value interface{}
}

type janitor struct {
	Interval time.Duration
	stop     chan bool
}

type backupData struct {
	data [2]string
	c    *bucketInternalPtr // when not nil a file compaction is requested
	file *os.File
}

type updateFunc func(k, v string) (string, string)


bucketInternal.go
package jac

import (
	"runtime"
	"time"
)

// back-end methods

// Return a declare bucketInternal with a given default expiration duration and cleanup
// interval. If the expiration duration is less than one (or NoExpiration),
// the items in the bucketInternal never expire (by default), and must be deleted
// manually. If the cleanup interval is less than one, expired items are not
// deleted from the bucketInternal before calling c.deleteExpired().
func declare(defaultExpiration, cleanupInterval time.Duration) *bucketInternalPtr {
	items := make(map[string]Item)
	return newCacheWithJanitor(defaultExpiration, cleanupInterval, items)
}

func (c *bucketInternal) set(k string, x interface{}, d time.Duration) {
	var e int64
	if d == DefaultExpiration {
		d = c.defaultExpiration
	}
	if d > 0 {
		e = time.Now().Add(d).UnixNano()
	}
	c.items[k] = Item{
		Object:     x,
		Expiration: e,
	}
}

func (c *bucketInternal) get(k string) (interface{}, bool) {
	item, found := c.items[k]
	if !found {
		return nil, false
	}
	// "Inlining" of expired
	if item.Expiration > 0 {
		if time.Now().UnixNano() > item.Expiration {
			return nil, false
		}
	}
	return item.Object, true
}

func (c *bucketInternal) delete(k string) (interface{}, bool) {
	if c.onEvicted != nil {
		if v, found := c.items[k]; found {
			delete(c.items, k)
			return v.Object, true
		}
	}
	delete(c.items, k)
	return nil, false
}

func (c *bucketInternal) deleteExpired() {
	var evictedItems []keyAndValue
	now := time.Now().UnixNano()
	c.mu.Lock()
	for k, v := range c.items {
		// "Inlining" of expired
		if v.Expiration > 0 && now > v.Expiration {
			ov, evicted := c.delete(k)
			if evicted {
				evictedItems = append(evictedItems, keyAndValue{k, ov})
			}
		}
	}
	c.mu.Unlock()
	for _, v := range evictedItems {
		c.onEvicted(v.key, v.value)
	}
}

func (c *bucketInternal) bucketItems() map[string]Item {
	c.mu.RLock()
	defer c.mu.RUnlock()
	m := make(map[string]Item, len(c.items))
	now := time.Now().UnixNano()
	for k, v := range c.items {
		// "Inlining" of expired
		if v.Expiration > 0 {
			if now > v.Expiration {
				continue
			}
		}
		m[k] = v
	}
	return m
}

func (j *janitor) run(c *bucketInternal) {
	ticker := time.NewTicker(j.Interval)
	for {
		select {
		case <-ticker.C:
			c.deleteExpired()
		case <-j.stop:
			ticker.Stop()
			return
		}
	}
}

func stopJanitor(c *bucketInternalPtr) {
	c.janitor.stop <- true
}

func runJanitor(c *bucketInternal, ci time.Duration) {
	j := &janitor{
		Interval: ci,
		stop:     make(chan bool),
	}
	c.janitor = j
	go j.run(c)
}

func newCache(de time.Duration, m map[string]Item) *bucketInternal {
	if de == 0 {
		de = -1
	}
	c := &bucketInternal{
		defaultExpiration: de,
		items:             m,
	}
	return c
}

func newCacheWithJanitor(de time.Duration, ci time.Duration, m map[string]Item) *bucketInternalPtr {
	c := newCache(de, m)
	// This trick ensures that the janitor goroutine (which--granted it
	// was enabled--is running deleteExpired on c forever) does not keep
	// the returned C object from being garbage collected. When it is
	// garbage collected, the finalizer stops the janitor goroutine, after
	// which c can be collected.
	C := &bucketInternalPtr{c}
	if ci > 0 {
		runJanitor(c, ci)
		runtime.SetFinalizer(C, stopJanitor)
	}
	return C
}

func (item Item) expired() bool {
	if item.Expiration == 0 {
		return false
	}
	return time.Now().UnixNano() > item.Expiration
}

func (c *Bucket) set(k, v string, t time.Duration, bck bool) {
	if k == "" {
		return
	}
	if c.writer != nil && bck {
		select {
		case c.writer <- backupData{
			data: [2]string{k, v},
			file: c.file,
		}:
		//default:
		case <-time.After(time.Duration(options.LoadDelayMs) * time.Millisecond):
			//println("lost")
		}
	}
	c.bucket.set(k, v, t)
}


internalProcesses.go
package jac

import (
	"encoding/json"
	"fmt"
	"time"
)

// working file write handler
func writeHandler(consolidateTimers map[string]int64) {
	defer func() {
		if r := recover(); r != nil {
			writeHandler(consolidateTimers)
		}
	}()
	if consolidateTimers == nil {
		consolidateTimers = make(map[string]int64)
	}
	if verbose {
		println("start writeHandler")
	}
	for {
		select {
		case <-writeRstChannel:
			if verbose {
				fmt.Println("writeHandler closing")
			}
			writeRstChannel <- nil
		case nw := <-writeChannel:
			//fmt.Println("received", nw)
			if nw.file != nil {
				if nw.c != nil {
					name := fmt.Sprintf("%v", nw.file.Name())
					tm, skip := consolidateTimers[name]
					if skip {
						skip = time.Now().Unix()-tm < int64(options.IntervalCompacting)
					} else {
						consolidateTimers[name] = time.Now().Unix() - 1
					}
					if !skip {
						// consolidation
						if e := nw.file.Truncate(0); e == nil {
							if _, e = nw.file.Seek(0, 0); e == nil {
								for i, v := range nw.c.bucketItems() {
									if fmt.Sprintf("%v", v.Object) != "" {
										if data, err := json.Marshal(FileData{
											Key:   i,
											Value: fmt.Sprintf("%v", v.Object),
										}); err == nil {
											_, _ = nw.file.WriteString(string(data) + "\n")
										}
									}
								}
							}
						}
						continue
					}
				}
				// Update
				if data, err := json.Marshal(FileData{
					Key:   nw.data[0],
					Value: nw.data[1],
				}); err == nil {
					_, _ = nw.file.WriteString(string(data) + "\n")
				}
			}
		}
	}
}

// working file compact handler
func compactHandler(c Bucket) {
	defer func() {
		if r := recover(); r != nil {
			compactHandler(c)
		}
	}()
	if verbose {
		println("start compactHandler for " + c.name)
	}
	for {
		select {
		case <-c.cr:
			if verbose {
				fmt.Println("compactHandler closing for", c.name)
			}
		case <-time.After(time.Duration(options.IntervalCompacting) * time.Second):
			// in case of a zombie
			if c.file == nil {
				return
			}
			c.bucket.deleteExpired()
			c.writer <- backupData{
				c:    c.bucket,
				file: c.file,
			}
		}
	}

}


cache.go
package jac

import (
	"bufio"
	"encoding/gob"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// Initialise prepare the cache for use. It accept two parameters:
//  v : set to true for verbose (useful for development purposes)
//  o : are the database options (see inm types.go for further details)
func Initialise(v bool, o *Options) error {

	verbose = v
	ex, err := os.Executable()
	if err != nil {
		return err
	}

	// set default values
	options = Options{
		ExpirationTime:     0,
		IntervalCompacting: 1440 * 60,
		InternalBuffering:  10,
		LoadDelayMs:        10,
		MaximumAge:         5 * 60,
		WorkingFolder:      filepath.Dir(ex) + "/",
		RecoveryFolder:     filepath.Dir(ex) + "/",
	}
	if o != nil {
		// read users values
		// 0 values are not allowed and will generate a non blocking error
		// values below a given minimum will be rejected
		err = IllegalParameter
		if o.MaximumAge > 0 {
			options.MaximumAge = o.MaximumAge
			err = nil
		}
		if o.IntervalCompacting >= 60 {
			options.IntervalCompacting = o.IntervalCompacting
			err = nil
		}
		if o.InternalBuffering >= 10 {
			options.InternalBuffering = o.InternalBuffering
			err = nil
		}
		if o.LoadDelayMs >= 5 {
			options.LoadDelayMs = o.LoadDelayMs
			err = nil
		}
		if o.ExpirationTime >= 0 {
			options.ExpirationTime = o.ExpirationTime
			err = nil
		}
		if o.WorkingFolder != "" {
			if err = os.MkdirAll(o.WorkingFolder, os.ModePerm); err != nil {
				fmt.Println(err)
				os.Exit(0)
			}
			options.WorkingFolder = o.WorkingFolder
			if options.WorkingFolder != "" && options.WorkingFolder[len(options.WorkingFolder)-1] != '/' {
				options.WorkingFolder += "/"
			}
		}
		if o.RecoveryFolder != "" {
			if err = os.MkdirAll(o.RecoveryFolder, os.ModePerm); err != nil {
				fmt.Println(err)
				os.Exit(0)
			}
			options.RecoveryFolder = o.RecoveryFolder
			if options.RecoveryFolder != "" && options.RecoveryFolder[len(options.RecoveryFolder)-1] != '/' {
				options.RecoveryFolder += "/"
			}
		}
	}

	// set internal processes and channels
	writeChannel = make(chan backupData, options.InternalBuffering)
	writeRstChannel = make(chan interface{})
	go writeHandler(nil)
	return nil
}

// Terminate closes the cache and relative processes
func Terminate() {
	writeRstChannel <- nil
	<-writeRstChannel
}

// NewBucket create a new bucket in the cache.
//  If a rec file or a data file are present and are not older than time.Now() - maxage,
//  they will be loaded in the cache
func NewBucket(name string, exp time.Duration) (c Bucket, e error) {
	c.name = name
	c.writer = writeChannel
	c.bucket = declare(time.Duration(options.ExpirationTime)*time.Second, time.Duration(2*options.ExpirationTime)*time.Second)
	c.cr = make(chan interface{})
	// first check for recovery file from normal termination
	statInfo, err := os.Stat(options.RecoveryFolder + name + ".rec")
	if !os.IsNotExist(err) {
		if (time.Now().Unix() - statInfo.ModTime().Unix()) < options.MaximumAge*60 {
			c.file, e = os.Create(options.WorkingFolder + name + ".data")
			if f, err := os.Open(options.RecoveryFolder + name + ".rec"); err == nil {
				var data map[string]Item
				dataDecoder := gob.NewDecoder(f)
				if err = dataDecoder.Decode(&data); err == nil {
					for i, v := range data {
						c.Set(i, fmt.Sprintf("%v", v.Object), exp, true)
					}
				} else {
					c.bucket = declare(time.Duration(options.ExpirationTime)*time.Second, time.Duration(2*options.ExpirationTime)*time.Second)
				}
				_ = f.Close()
			}
		}
		if e = os.Remove(options.RecoveryFolder + name + ".rec"); err != nil {
			return
		}
		go compactHandler(c)
		return
	}
	// check for working file remnants from server crash
	statInfo, err = os.Stat(options.WorkingFolder + name + ".data")
	if !os.IsNotExist(err) {
		if (time.Now().Unix() - statInfo.ModTime().Unix()) < options.MaximumAge*60 {

			if c.file, err = os.Open(options.WorkingFolder + name + ".data"); err == nil {
				scanner := bufio.NewScanner(c.file)
				for scanner.Scan() {
					var entry FileData
					if err := json.Unmarshal([]byte(scanner.Text()), &entry); err == nil {
						c.Update(entry.Key, entry.Value, exp, true)
					}
				}
				_ = c.file.Close()
			}
		}
		go compactHandler(c)
		return
	}
	c.file, e = os.Create(options.WorkingFolder + name + ".data")
	go compactHandler(c)
	return
}

// Close closes a bucket storing values in the recovery data
//  if keep is false the working data file will be deleted
func (c *Bucket) Close(keep bool) {
	if f, err := os.Create(options.RecoveryFolder + c.name + ".rec"); err == nil {
		// serialize the data
		dataEncoder := gob.NewEncoder(f)
		err = dataEncoder.Encode(c.bucket.bucketItems())
		_ = c.file.Close()
		_ = f.Close()
		if err == nil {
			if !keep {
				_ = os.Remove(options.WorkingFolder + c.name + ".data")
			}
		} else {
			_ = os.Remove(options.RecoveryFolder + c.name + ".rec")
		}
	} else {
		_ = c.file.Close()
	}
	c.file = nil
	go func() { c.cr <- nil }()
}

// Get read the value associated to the key k
//  It also returns false if the key has not value associated to it
func (c *Bucket) Get(k string) (string, bool) {
	c.bucket.mu.RLock()
	item, found := c.bucket.items[k]
	if !found {
		c.bucket.mu.RUnlock()
		return "", false
	}
	if item.Expiration > 0 {
		if time.Now().UnixNano() > item.Expiration {
			c.bucket.mu.RUnlock()
			return "", false
		}
	}
	c.bucket.mu.RUnlock()
	v := fmt.Sprintf("%v", item.Object)
	return v, found || v != ""
}

// GetWithExpiration performs the same operation as Get but it also returns the
//  value expiration time
func (c *Bucket) GetWithExpiration(k string) (string, time.Time, bool) {
	c.bucket.mu.RLock()
	item, found := c.bucket.items[k]
	if !found {
		c.bucket.mu.RUnlock()
		return "", time.Time{}, false
	}

	if item.Expiration > 0 {
		if time.Now().UnixNano() > item.Expiration {
			c.bucket.mu.RUnlock()
			return "", time.Time{}, false
		}

		// Return the item and the expiration time
		c.bucket.mu.RUnlock()
		return fmt.Sprintf("%v", item.Object), time.Unix(0, item.Expiration), true
	}

	// If expiration <= 0 (i.e. no expiration time set) then return the item
	// and a zeroed time.Time
	c.bucket.mu.RUnlock()
	return fmt.Sprintf("%v", item.Object), time.Time{}, true
}

// Set writes a new key/value pair with a given expiration time t and
//  It marks the key/value pair persistent if pers is true.
func (c *Bucket) Set(k string, vn interface{}, t time.Duration, pers bool) {
	if k == "" || vn == nil {
		return
	}
	v := anything2String(vn)
	if c.writer != nil && pers {
		select {
		case c.writer <- backupData{
			data: [2]string{k, v},
			file: c.file,
		}:
		case <-time.After(time.Duration(options.LoadDelayMs) * time.Millisecond):
		}
	}

	var e int64
	if t == DefaultExpiration {
		t = c.bucket.defaultExpiration
	}
	if t > 0 {
		e = time.Now().Add(t).UnixNano()
	}
	c.bucket.mu.Lock()
	c.bucket.items[k] = Item{
		Object:     v,
		Expiration: e,
	}
	c.bucket.mu.Unlock()
}

// Update updates the key/value pair with a given expiration time t and
//  It marks the key/value pair persistent if pers is true. The working file is not changed
//  even if pers is true
func (c *Bucket) Update(k string, vn interface{}, t time.Duration, pers bool) {
	c.bucket.mu.Lock()
	if k == "" || vn == nil {
		return
	}
	v := anything2String(vn)
	if _, found := c.bucket.get(k); found {
		c.bucket.set(k, v, t)
	} else {
		c.set(k, v, t, pers)
	}
	c.bucket.mu.Unlock()
}

// Replace replaces an existing key/value pair only it already existing.
//  It marks the key/value pair persistent if pers is true.
func (c *Bucket) Replace(k string, vn interface{}, t time.Duration, pers bool) {
	c.bucket.mu.Lock()
	if k == "" || vn == nil {
		return
	}
	v := anything2String(vn)
	if _, found := c.bucket.get(k); found {
		c.set(k, v, t, pers)
	}
	c.bucket.mu.Unlock()
}

// Add add a new key/value pair only if it does not already
//  It marks the key/value pair persistent if pers is true.
func (c *Bucket) Add(k string, vn interface{}, t time.Duration, pers bool) (string, bool) {
	c.bucket.mu.Lock()
	defer c.bucket.mu.Unlock()
	if k == "" || vn == nil {
		return "", false
	}
	v := anything2String(vn)
	if val, found := c.bucket.get(k); found && val != "" {
		return fmt.Sprintf("%v", val), true
	} else {
		c.set(k, v, t, pers)
		return "", false
	}
}

// FunctionUpdate updates the new key/value with a custom function of type
//  func(k, v string) (string, string). The function is given the actual key/value pair is present
//  and key/"" otherwise and it expects a new key/value pair.
//  if the function modofies the key, the old key/pair is deleted.
//  It marks the key/value pair persistent if pers is true.
func (c *Bucket) FunctionUpdate(k string, f updateFunc, t time.Duration, pers bool) (string, string, bool) {
	c.bucket.mu.Lock()
	defer c.bucket.mu.Unlock()
	if k == "" {
		return "", "", false
	}
	if val, found := c.bucket.get(k); found && val != nil {
		newK, newV := f(k, fmt.Sprintf("%v", val))
		if newK != k {
			c.bucket.delete(k)
		}
		c.set(newK, newV, t, pers)
		return newK, newV, true
	} else {
		newK, newV := f(k, "")
		c.set(newK, newV, t, pers)
		return newK, newV, false
	}
}

// Items returns all elements in the bucket as a map[string]string
func (c *Bucket) Items() (rt map[string]string) {
	rt = make(map[string]string)
	for i, v := range c.bucket.bucketItems() {
		if val := fmt.Sprintf("%v", v.Object); val != "" {
			rt[i] = val
		}
	}
	return
}

// Delete permanently removes an item from the bucket
func (c *Bucket) Delete(k string) {
	c.bucket.mu.Lock()
	v, evicted := c.bucket.delete(k)
	c.bucket.mu.Unlock()
	if evicted {
		c.bucket.onEvicted(k, v)
	}
}

// Compact initiate a compation request of the working files.
//  Please note that all values will be used and any previously indicated persistence flag will be ignored.
func (c *Bucket) Compact() {
	c.bucket.deleteExpired()
	select {
	case c.writer <- backupData{
		data: [2]string{},
		c:    c.bucket,
		file: c.file,
	}:
	default:
	}
}

// Flush deletes all items from the bucket.
func (c *Bucket) Flush() {
	c.bucket.mu.Lock()
	c.bucket.items = map[string]Item{}
	c.bucket.mu.Unlock()
}

// ItemCount returns the number of items in the bucket. This may include items that have
//  expired, but have not yet been cleaned up.
func (c *Bucket) ItemCount() int {
	c.bucket.mu.RLock()
	n := len(c.bucket.items)
	c.bucket.mu.RUnlock()
	return n
}

// OnEvicted sets an (optional) function that is called with the key and value when an
//  item is evicted from the bucketInternal. (Including when it is deleted manually, but
//  not when it is overwritten.) Set to nil to disable.
func (c *Bucket) OnEvicted(f func(string, interface{})) {
	c.bucket.mu.Lock()
	c.bucket.onEvicted = f
	c.bucket.mu.Unlock()
}

// DeleteExpired deletes all expired items from the bucketInternal.
func (c *Bucket) DeleteExpired() {
	c.bucket.deleteExpired()
}
